### PR TITLE
QML: Expose library track table as a listmodel

### PR DESCRIPTION
**Pull request by Holzhaus**
This is a first implementation to make the library usable in QML.

Actually I'm unsure if using a listmodel works for QML after all. For example, I don't really know how to implement multiple columns that may even be reorderable. But it works fine for testing and as an initial implementation.

After we did some testing we may have to go with a tablemodel after all. This should not block this PR though.

### CODE REVIEW

**Comment by Holzhaus** [comment]
@Be-ing did you try this out already?

**Comment by uklotzde** [comment]
Thank your for providing this template as a starting point! I'll try if I can adapt and plug-in my aoide model.

**Comment by Holzhaus** [comment]
> Thank your for providing this template as a starting point! I'll try if I can adapt and plug-in my aoide model.

Before spending a lot of time on that we should evaluate if this is indeed how our model should look. This approach makes it impossible to use an old-school table with reorderable/resizable columns.

I'd suggest to merge this and then try to implement a better library view in QML.

A possibility would be to add another layer and implement a custom wrapper that maps a listmodel back to a table model.

Something like:

```qml

TableView {
    model: CustomListTableModel {
        model: Mixxx.Library.model

        CustomListTableModelColumn { role: "title" } 
        CustomListTableModelColumn { role: "artist" }
        CustomListTableModelColumn { role: "album" } 
        CustomListTableModelColumn { role: "albumartist" } 
    } 
} 
```

This would make sure that any the default order of columns is a property of the UI, not the underlying data model. It would also ensure that only columns are shown that the skin designer thought about and provides a proper delegate for. 
But I don't know if it's a good idea performance-wise.

**Comment by uklotzde** [comment]
Performance should not be an issue for the few items that are actually visible. They are reusing the model items behind the scenes. And when we stop loading the whole database at once into the model this should become even less of an issue.

**Comment by Be-ing** [comment]
> @Be-ing did you try this out already?

Not yet. I will soon.

**Comment by Be-ing** [comment]
> A possibility would be to add another layer and implement a custom wrapper that maps a listmodel back to a table model.

I think it would be better to make a custom delegate for a list that mimics a table. We can try both.

File: res/skins/QMLDemo/Library.qml [hunk 1]
```diff
@@ -1,20 +1,56 @@
...
+                }
+
+                MouseArea {
+                    id: dragArea
+
+                    anchors.fill: parent
+                    drag.target: dragItem
+                    onPressed: parent.grabToImage(function(result) {
+                        dragItem.Drag.imageSource = result.url;
+                    })
```
Comment (Swiftb0y) [inline res/skins/QMLDemo/Library.qml:49]:
```suggestion
                    onPressed: parent.grabToImage(result => {
                        dragItem.Drag.imageSource = result.url;
                    })
```

File: res/skins/QMLDemo/Mixxx/PlayerDropArea.qml [hunk 2]
```diff
@@ -15,12 +15,15 @@ DropArea {
 
             console.log("Drag from group " + sourceGroup);
             player.cloneFromGroup(sourceGroup);
+            drop.accepted = true;
```
Comment (Swiftb0y) [inline res/skins/QMLDemo/Mixxx/PlayerDropArea.qml:18]: this doesn't seems to be strictly relevant to this PR? Could you extract it into its own PR?

File: res/skins/QMLDemo/Mixxx/PlayerDropArea.qml [hunk 3]
```diff
@@ -15,12 +15,15 @@ DropArea {
...
             return ;
         }
         if (drop.hasUrls) {
             let url = drop.urls[0];
             console.log("Dropped URL '" + url + "' on deck " + group);
             player.loadTrackFromLocationUrl(url);
+            drop.accepted = true;
+            return ;
```
Comment (Swiftb0y) [inline res/skins/QMLDemo/Mixxx/PlayerDropArea.qml:26]: same as above

File: src/skin/qml/qmllibraryproxy.cpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+QmlLibraryProxy::QmlLibraryProxy(Library* pLibrary, QObject* parent)
+        : QObject(parent),
+          m_pLibrary(pLibrary),
+          m_pModel(
+                  new QmlLibraryTrackListModel(m_pLibrary->trackTableModel())) {
+}
+
+QmlLibraryProxy::~QmlLibraryProxy() {
+    delete m_pModel;
+}
```
Comment (Swiftb0y) [inline src/skin/qml/qmllibraryproxy.cpp:21]: parented_ptr?

File: src/skin/qml/qmllibraryproxy.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace mixxx {
+namespace skin {
+namespace qml {
+
+class QmlLibraryTrackListModel;
+
+class QmlLibraryProxy : public QObject {
+    Q_OBJECT
+    Q_PROPERTY(mixxx::skin::qml::QmlLibraryTrackListModel* model MEMBER m_pModel CONSTANT)
```
Comment (Swiftb0y) [inline src/skin/qml/qmllibraryproxy.h:14]: from https://doc.qt.io/qt-5/properties.html:
> Here is an example showing how to export member variables as Qt properties using the MEMBER keyword. Note that a NOTIFY signal must be specified to allow QML property bindings.

I guess that is untrue when the property is also marked CONSTANT? since
> The presence of the CONSTANT attribute indicates that the property value is constant. For a given object instance, the READ method of a constant property must return the same value every time it is called. This constant value may be different for different instances of the object. A constant property cannot have a WRITE method or a NOTIFY signal.

File: src/skin/qml/qmlskin.cpp [hunk 6]
```diff
@@ -204,6 +206,27 @@ QWidget* QmlSkin::loadSkin(QWidget* pParent,
                         return pConfigProxy;
                     }));
 
+    qmlRegisterUncreatableType<QmlLibraryTrackListModel>("Mixxx",
+            0,
+            1,
+            "LibraryTrackListModel",
+            "LibraryTrackListModel objects can't be created directly, "
+            "please use Mixxx.Library.model");
```
Comment (Swiftb0y) [inline src/skin/qml/qmlskin.cpp:214]: Is there a specific reason why we are making it uncreatable?

**Review by Swiftb0y — CHANGES_REQUESTED** [review]
preliminary review. wasn't able to look at qmllibrarytracklistmodel in detail yet.

File: src/skin/qml/qmlskin.cpp (hunk 6 above)
Comment (Holzhaus) [reply src/skin/qml/qmlskin.cpp:214]: Yes. You need to pass the actual library tablemodel to the constructor, which is not possible in qml. We could implement lazy setting, but that is more complex, requires exposing the underlying library track model to and I'm unclear what the benefit would be.

File: src/skin/qml/qmllibraryproxy.cpp (hunk 4 above)
Comment (Holzhaus) [reply src/skin/qml/qmllibraryproxy.cpp:21]: I don't think that works with member properties.

File: res/skins/QMLDemo/Library.qml (hunk 1 above)
Comment (Holzhaus) [reply res/skins/QMLDemo/Library.qml:49]: I think this is less readable but ok.

File: res/skins/QMLDemo/Mixxx/PlayerDropArea.qml (hunk 2 above)
Comment (Holzhaus) [reply res/skins/QMLDemo/Mixxx/PlayerDropArea.qml:18]: No strictly, but it's relevant since the qml library implement drag and drop.

File: src/skin/qml/qmllibraryproxy.h (hunk 5 above)
Comment (Holzhaus) [reply src/skin/qml/qmllibraryproxy.h:14]: Yes, I think so. It works.

File: res/skins/QMLDemo/Mixxx/PlayerDropArea.qml (hunk 2 above)
Comment (Holzhaus) [reply res/skins/QMLDemo/Mixxx/PlayerDropArea.qml]: https://github.com/mixxxdj/mixxx/pull/4067

File: src/skin/qml/qmlskin.cpp (hunk 6 above)
Comment (Swiftb0y) [reply src/skin/qml/qmlskin.cpp:214]: The benefit would be that it would be more aligned with how the models are usually done in QML, but because of the technical challenges you just described, it obviously doesn't make much sense. The current solution is fine for me.

File: src/skin/qml/qmllibraryproxy.cpp (hunk 4 above)
Comment (Swiftb0y) [reply src/skin/qml/qmllibraryproxy.cpp:21]: not sure, I just thought it would suit the usecase of parented_ptr but I don't really understand how it works or how to use it. I guess @daschuer should know.

File: src/skin/qml/qmllibraryproxy.cpp [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+#include "skin/qml/qmllibrarytracklistmodel.h"
+
+namespace mixxx {
+namespace skin {
+namespace qml {
+
+QmlLibraryProxy::QmlLibraryProxy(Library* pLibrary, QObject* parent)
+        : QObject(parent),
+          m_pLibrary(pLibrary),
+          m_pModel(
```
Comment (uklotzde) [inline src/skin/qml/qmllibraryproxy.cpp:15]: Why not pass `this` as `parent` and use `parented_ptr` for the member as suggested? Object hierarchies are what Qt is supposed to manage by parenting.

File: src/skin/qml/qmllibrarytracklistmodel.cpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        {QmlLibraryTrackListModel::ArtistRole, "artist"},
+        {QmlLibraryTrackListModel::AlbumRole, "album"},
+        {QmlLibraryTrackListModel::AlbumArtistRole, "albumArtist"},
+        {QmlLibraryTrackListModel::FileUrlRole, "fileUrl"},
+};
+}
+
+QVariant QmlLibraryTrackListModel::data(const QModelIndex& proxyIndex, int role) const {
+    if (!proxyIndex.isValid()) {
+        return QVariant();
```
Comment (uklotzde) [inline src/skin/qml/qmllibrarytracklistmodel.cpp:18]: Let's simplify all occurrences of `return QVariant();` to `return {};`

File: src/skin/qml/qmllibrarytracklistmodel.cpp [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    if (column < 0) {
+        return QVariant();
+    }
+
+    return QIdentityProxyModel::data(proxyIndex.siblingAtColumn(column), Qt::DisplayRole);
+}
+
+int QmlLibraryTrackListModel::columnCount(const QModelIndex& parent) const {
+    // This is a list model, i.e. entries without a parent have exactly one column.
+    if (!parent.isValid()) {
```
Comment (uklotzde) [inline src/skin/qml/qmllibrarytracklistmodel.cpp:71]: VERIFY_OR_DEBUG_ASSERT(!parent.isValid()) {
// ...
return 0;
}
// ...
return 1;

File: src/skin/qml/qmllibrarytracklistmodel.h [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    enum Roles {
+        TitleRole = Qt::UserRole,
+        ArtistRole,
+        AlbumRole,
+        AlbumArtistRole,
+        FileUrlRole,
+    };
+    Q_ENUM(Roles);
+
+    QmlLibraryTrackListModel(LibraryTableModel* pModel, QObject* pParent = nullptr)
```
Comment (uklotzde) [inline src/skin/qml/qmllibrarytracklistmodel.h:21]: librarytablemodel.h does not need to be included when moving this implementation into the .cpp file.

File: src/skin/qml/qmllibraryproxy.cpp (hunk 7 above)
Comment (Holzhaus) [reply src/skin/qml/qmllibraryproxy.cpp]: I thought this wouldn't work due to the Q_PROPERTY, but it does :tada: eff29a67989b95e45b0c5de5faaa8fb1dfe4a6f3

File: src/skin/qml/qmllibrarytracklistmodel.cpp (hunk 8 above)
Comment (Holzhaus) [reply src/skin/qml/qmllibrarytracklistmodel.cpp]: 792acbc7dad1e6e1b0132c39e0ddb49d610a5225

File: src/skin/qml/qmllibrarytracklistmodel.cpp (hunk 9 above)
Comment (Holzhaus) [reply src/skin/qml/qmllibrarytracklistmodel.cpp]: 32da274627f93acdd73fa87c5d559783c9cb5121

File: src/skin/qml/qmllibrarytracklistmodel.h (hunk 10 above)
Comment (Holzhaus) [reply src/skin/qml/qmllibrarytracklistmodel.h]: 07b2ec088973dbac277851559f7eb1c1300e53f8

File: src/skin/qml/qmllibraryproxy.cpp (hunk 4 above)
Comment (Holzhaus) [reply src/skin/qml/qmllibraryproxy.cpp]: I just tried it and it works: https://github.com/mixxxdj/mixxx/commit/eff29a67989b95e45b0c5de5faaa8fb1dfe4a6f3

**Review by uklotzde — APPROVED** [review]

**Comment by Be-ing** [comment]
I tested this briefly and it is a nice start to make the QML more usable for testing.

File: src/skin/qml/qmlskin.cpp (hunk 6 above)
Comment (Be-ing) [reply src/skin/qml/qmlskin.cpp:214]: Please add a comment briefly explaining the technical challenges.

Comment (Holzhaus) [reply src/skin/qml/qmlskin.cpp:223]: cafb002a517f5b8752496c3f25f296a50470c2fb

**Comment by Holzhaus** [comment]
Merge?

**Review by Swiftb0y — APPROVED** [review]
We have one thorough review, one QML review and three different people testing so this LGTM.

File: src/skin/qml/qmlskin.cpp [hunk 11]
```diff
@@ -132,6 +134,15 @@ QWidget* QmlSkin::loadSkin(QWidget* pParent,
     qmlRegisterType<QmlControlProxy>("Mixxx", 0, 1, "ControlProxy");
     qmlRegisterType<QmlWaveformOverview>("Mixxx", 0, 1, "WaveformOverview");
 
+    // Any uncreateable non-singletom types registered here require arguments
```
Comment (Swiftb0y) [inline src/skin/qml/qmlskin.cpp:137]:
```suggestion
    // Any uncreateable non-singleton types registered here require arguments
```
typo

Comment (Holzhaus) [reply src/skin/qml/qmlskin.cpp]: thx, I amended the comment.

**Comment by Holzhaus** [comment]
Merge?
